lib/router/src/cache.rs

```diff
@@ -1,426 +1,476 @@
+    pub(crate) value: T,
 }
 
-/// A smart pointer that updates an access time when dropped.
-///
-/// Wraps a mutable reference to a `V`-typed value.
-///
-/// When the guard is dropped, the value's `last_access` time is updated with the provided
-/// time source.
-#[derive(Debug)]
-pub struct Access<'a, T: 'a, N: Now + 'a = ()> {
-    node: &'a mut Node<T>,
-    now: &'a N,
+#[derive(Debug, PartialEq)]
+pub struct CapacityExhausted {
+    pub(crate) capacity: usize,
 }
 
-/// A handle to a `Cache` that has capacity for at least one additional value.
-#[derive(Debug)]
-pub struct Reserve<'a, K: Hash + Eq + 'a, V: 'a, N: 'a> {
+/// A handle to a cache that has capacity for at least one additional value.
+pub struct Reserve<'a, K: Clone + Eq + Hash, V> {
+    expirations: &'a mut DelayQueue<K>,
+    expires: Duration,
     vals: &'a mut IndexMap<K, Node<V>>,
-    now: &'a N,
-}
-
-#[derive(Clone, Debug, PartialEq)]
-pub struct CapacityExhausted {
-    pub capacity: usize,
 }
 
 // ===== impl Cache =====
 
-impl<K: Hash + Eq, V> Cache<K, V, ()> {
-    pub fn new(capacity: usize, max_idle_age: Duration) -> Self {
-        Self {
+impl<K: Clone + Eq + Hash, V> Cache<K, V> {
+    pub fn new(capacity: usize, expires: Duration) -> (Lock<Cache<K, V>>, PurgeCache<K, V>) {
+        assert!(capacity != 0);
+        let cache = Self {
             capacity,
+            expires,
+            expirations: DelayQueue::with_capacity(capacity),
             vals: IndexMap::default(),
-            max_idle_age,
-            now: (),
-        }
+        };
+        let cache = Lock::new(cache);
+        let bg_purge = PurgeCache {
+            cache: cache.clone(),
+            interval: Interval::new_interval(expires),
+        };
+
+        (cache, bg_purge)
+    }
+
+    pub fn access(&mut self, key: &K) -> Option<Access<K, V>> {
+        let node = self.vals.get_mut(key)?;
+        Some(Access {
+            expires: self.expires,
+            expirations: &mut self.expirations,
+            node,
+        })
     }
-}
 
-impl<K: Hash + Eq, V, N: Now> Cache<K, V, N> {
-    /// Accesses a route.
-    ///
-    /// A mutable reference to the route is wrapped in the returned `Access` to
-    /// ensure that the access-time is updated when the reference is released.
-    pub fn access<Q>(&mut self, key: &Q) -> Option<Access<V, N>>
-    where
-        Q: Hash + Equivalent<K>,
-    {
-        let v = self.vals.get_mut(key)?;
-        Some(v.access(&self.now))
+    pub fn capacity(&self) -> usize {
+        self.capacity
     }
 
-    /// Ensures that there is capacity to store an additional route.
-    ///
-    /// Returns a handle that may be used to store an ite,. If there is no available
-    /// capacity, idle entries may be evicted to create capacity.
-    ///
-    /// An error is returned if there is no available capacity.
-    pub fn reserve(&mut self) -> Result<Reserve<K, V, N>, CapacityExhausted> {
+    pub fn reserve(&mut self) -> Async<Reserve<K, V>> {
```

Comment:
What if we changed this from a function returning `Async` to a function returning a `Future`, that returns `Ok(Async::Ready(Reserve<K, V>))` when we're able to reserve an entry? I think that could move some logic out of the complex state machine in `Router::call`

Reply:
Since we want to fail the request when capacity is exhausted (see https://github.com/linkerd/linkerd2-proxy/pull/247#discussion_r293508066), this should probably return `Result<Reserve<...>, CapacityExhausted>` rather than `Async`, as it did previously.